elife/multiservice.sls

```diff
@@ -0,0 +1,55 @@
+
+# systemd managed group processes
+
+# 1. N number of the given process are running
+# 2. the pool of processes can be grown or shrunk
+# 3. broken processes fail highstate as they ordinarily would
+
+{% for process, opts in pillar.elife.multiservice.services.items() %}
+    {% set num_processes = opts["num_processes"] %}
+
+# ensure the target state is available and running
+{{ process }}-controller.target:
+    file.managed:
+        - name: /lib/systemd/system/{{ process }}-controller.target
+        - source: salt://elife/templates/process-controller.target
+
+    service.running:
+        - enable: true
+        - require:
+            - file: {{ process }}-controller.target
+
+# ensure precisely N processes are enabled. restart controller if N has changed
+# enabled processes are started when the controller state changes
+{{ process }}-set-enabled:
+    cmd.run:
+        - name: |
+            set -e
+            systemctl stop {{ process }}@{0..99} # stop any running instances of this service.
+            systemctl disable {{ process }}@ # disable *everything*. implicit reload
+            {% if num_processes > 0 %}
+            systemctl enable {{ process }}@{0..{{ num_processes - 1}}} # enable just the range we're after. implicit reload
+            {% endif %}
+        - require:
+            - file: {{ opts["service_template"] }} # name of state that manages the systemd service template file
+        - onlyif:
+            # only run command when the expected number of processes (enabled and running) is not equal to the intended 
+            # number of processes. 
+            # use 'is-enabled' to find enabled units. it conveniently returns 'disabled' if all are disabled, so grep for 'enabled'
+            # use 'show' and grep for the pid for running units. it conveniently returns MainPID=0 for enabled-but-not-running units
+            - test {{ num_processes }} != $(systemctl is-enabled {{ process }}@{0..99} | grep 'enabled' | wc -l) || \
+              test {{ num_processes }} != $(systemctl show {{ process }}@{0..99} | grep '^MainPID=[^0]' | wc -l)
+        - watch_in:
+            - service: {{ process }}-controller.target
+
+# ensure N processes are running 
+{% for i in range(0, num_processes) %}
+{{ process }}@{{ i }}:
+    service.running:
+        - enable: true # redundant, but can't hurt
+        - require:
+            - {{ process }}-set-enabled
+            - {{ process }}-controller.target
```

Comment:
trying this out locally in combination with https://github.com/elifesciences/journal-cms-formula/pull/54 and the 18.04 Vagrant box. Main concern is that during a new build all processes are restarted, I don't see states or `systemctl` commands that accomplish that :thinking:

Reply:
Can't get it to work quickly, better to test it with the `basebox` one.

Reply:
Confirmed: processes are not restarted on a new highstate.

Reply:
Solved by using an explicit restart of the target.

Reply:
> Confirmed: processes are not restarted on a new highstate

they're not supposed to! 

these are regular processes that shouldn't be restarted if it's not necessary. if a project wants it's  set of processes restarted once a day on highstate then it can go above and beyond and issue a `systemctl restart myservice-controller` command.

Reply:
I really think we're on two different wavelengths here. I'm not going to merge the code with that restart statement in there until I understand and agree with what you're trying to do.

Reply:
It's not restarting on a daily highstate that is interesting, but during a deployment.

Existing projects that contain long-running processes rely on the Upstart script [etc-init-multiple-processes-parallel.conf](https://github.com/search?q=org%3Aelifesciences+%22etc-init-multiple-processes-parallel%22&type=Code) to restart all of these processes on each deployment or daily highstate. The one that don't [have workarounds to still restart them all the time](https://github.com/elifesciences/lax-formula/blob/master/salt/lax/bot-lax-adaptor.sls#L175-L177).

It is possible in theory to write enough `watch` statements to trigger restarts only when needed, but in practice it is too complex for projects to implement, so they resort to always restart. The recent case was rotating credentials, with services remaining on the old credentials for a while?

Reply:
Long story short: I am porting the current behavior of the Upstart implementation.

Reply:
ok, so these processes *are* being restarted on highstate, but only as a side effect because we're not being strict enough in writing watches for things like credentials and configuration changing?

I think this is a lame solution for broken formulas. If all it needs is a restart then there was no need to split the `{{ process }}-set` state into two or remove the `service.running` states. In fact, if we settle on a restart on every highstate and don't care that commands are being unnecessarily run then this whole file could be gutted and replaced with something simpler.

Reply:
take a look at this PR: https://github.com/elifesciences/builder-base-formula/pull/239

same effect, much simpler, better checks against running processes